Calculate balance based on pending txs

danfinlay opened the pull request:
Fixes #1746 by creating a new state object for tracking a computed balance that accounts for pending transactions.

Adds two new controllers:
- `balances.js`: An observable store that watches `ethStore`, `transactions`, and `networkChanged` and triggers an `update` among its children.
- `balance.js`: An observable store representing a single account and its balances. Currently only tracks `ethBalance`, but is designed to be extendable, and potentially track token balances in the future.

Adds the new controller state to `state.metamask.computedBalances`, which looks like this:
```
{
  "metamask: {
    "computedBalances": {
      "0xfdea65c8e26263f6d9a1b5de9555d2931a33b825": {
        "ethBalance": "0x0"
      }
  }
}
```

Updated the current UI to use this new computed balance in two places:
- The transaction approval view, which in turn disallows sending more than you will have after current pending transactions complete.
- The main account detail view.

danfinlay committed:
Add test template

danfinlay committed:
Add first passing balance calc test

danfinlay committed:
Check balances in parallel

danfinlay committed:
test not passing

danfinlay committed:
Add basic test for valueFor

danfinlay committed:
Fix valueFor test

danfinlay committed:
Add constructor comment

danfinlay committed:
Fix balance calc test

danfinlay committed:
Remove logs

danfinlay committed:
Add test for computing gas price

danfinlay committed:
Make tx calculations account for gas prices

danfinlay committed:
Improve test name

danfinlay committed:
Linted

danfinlay committed:
Merge branch 'i1746-BalanceBasedOnPending' into AddBalanceController

danfinlay committed:
Some progress

danfinlay committed:
Add state-labeled events to allow subscribing to any transaction's state change

danfinlay committed:
Got pending balance updating correctly

danfinlay committed:
Use computed balance for tx confirmation

danfinlay committed:
Add computed balance to account detail view

danfinlay committed:
Merge branch 'master' into AddBalanceController

danfinlay committed:
Fix test

flyswatter commented:
When I first load MM, it doesn't load the balance until the first block is detected. I'm starting to think that's a problem with ethStore...

kumavis commented on app/scripts/controllers/balance.js:
```diff
@@ -0,0 +1,60 @@
+    this.store = new ObservableStore(initState)
+
+    this.balanceCalc = new PendingBalanceCalculator({
+      getBalance: () => Promise.resolve(this._getBalance()),
```
idea: instead of wrapping with `Promise.resolve` you could make the fn async

kumavis commented on app/scripts/controllers/balance.js:
```diff
@@ -0,0 +1,60 @@
+    this.txController.on('submitted', update)
+    this.txController.on('confirmed', update)
+    this.txController.on('failed', update)
+    this.txController.blockTracker.on('block', update)
```
instead of 
```
// ...
this.txController.on('submitted', update)
this.txController.blockTracker.on('block', update)
```
you could just subscribe to `ethStore` and  `this.txController.memStore`

flyswatter replied:
I'm ok w/ the ethStore subscription instead of the blockTracker, but subscribing to specific, relevant txController state changes seems like a more efficient way to pre-filter the data we process, instead of processing on every txController mem change, which includes every single tx update of every sort.

kumavis commented on app/scripts/controllers/balance.js:
```diff
@@ -0,0 +1,60 @@
+  }
+
+  _getBalance () {
+    const store = this.ethStore.getState()
```
i like this style recently
```js
 const { accounts } = this.ethStore.getState()
```

flyswatter replied:
I'm slightly ambivalent about this, because it destructures and calls a method on the same line, so if there was an error on this line, it would be ambiguous which operation caused it.

kumavis commented on app/scripts/controllers/balance.js:
```diff
@@ -0,0 +1,60 @@
+      status: 'submitted',
+      err: undefined,
+    })
+    return Promise.resolve(pending)
```
might as well make the fn async 👍

danfinlay committed:
Remove dead code from eth-store

danfinlay committed:
Refactor eth-store into account-tracker

EthStore was only being used for tracking account balances and nonces now, so I removed its block-tracking duties, renamed it account-tracker, and removed it as a dependency from `KeyringController`, so that KRC can go live on without a hard dep on it.

danfinlay committed:
Fix computed-balances controller reference

danfinlay committed:
Merge branch 'master' into AddBalanceController

danfinlay committed:
Fix account-tracker references

danfinlay committed:
Fix accont-tracker merge bug

danfinlay committed:
Fix test references

danfinlay committed:
Made some requested changes

FlySwatter committed:
Merge branch 'master' into AddBalanceController

kumavis commented on test/unit/pending-balance-test.js:
```diff
@@ -0,0 +1,99 @@
+
+      const balanceCalculator = generateBalanceCalcWith([], zeroBn)
+      const result = balanceCalculator.valueFor(pendingTxs[0])
+      assert.equal(result.toString(), '6', 'computes one ether')
```
i dont understand how these values match `'6', 'computes one ether'`

flyswatter replied:
This is incorrectly labeled, should be "computes 6 wei of gas".

kumavis commented on test/unit/pending-balance-test.js:
```diff
@@ -0,0 +1,99 @@
+})
+
+function generateBalanceCalcWith (transactions, providerStub = zeroBn) {
+  const getPendingTransactions = () => Promise.resolve(transactions)
```
fyi can also be written as `const getPendingTransactions = async () => transactions`

kumavis commented on test/unit/pending-balance-test.js:
```diff
@@ -0,0 +1,99 @@
+  providerResultStub.result = providerStub
+  const provider = {
+    sendAsync: (_, cb) => { cb(undefined, providerResultStub) },
+    _blockTracker: {
```
is `provider._blockTracker.getCurrentBlock` being used? if so we should accept a blockTracker in addition to a provider, and not use the blockTracker thats on the provider

kumavis commented on app/scripts/lib/pending-balance-calculator.js:
```diff
@@ -0,0 +1,53 @@
+      this.getPendingTransactions(),
+    ])
+
+    const balance = results[0]
```
can also be written:
`const [ balance, pending ] = results`

kumavis commented on app/scripts/metamask-controller.js:
```diff
@@ -4,7 +4,7 @@ const promiseToCallback = require('promise-to-callback')
 const Dnode = require('dnode')
 const ObservableStore = require('obs-store')
-const EthStore = require('./lib/eth-store')
+const AccountTracker = require('./lib/account-tracker')
```
:+1:

kumavis reviewed (changes requested):
theres a lot going on here, lots of new code
hard to be sure I fully understand the behavior and side effects here, but generally looks good

would be nice to review with you over a call  @flyswatter

kumavis commented on app/scripts/controllers/balance.js:
```diff
@@ -0,0 +1,60 @@
+    })
+  }
+
+  registerUpdates () {
```
seems like this should be a private method `_registerUpdates`

kumavis commented on app/scripts/lib/pending-balance-calculator.js:
```diff
@@ -0,0 +1,53 @@
+    return `0x${balance.sub(pendingValue).toString(16)}`
+  }
+
+  valueFor (tx) {
```
consider renaming to `calculateMaxCost`

danfinlay committed:
Implemented feedback

danfinlay committed:
Merge branch 'AddBalanceController' of github.com:MetaMask/metamask-plugin into AddBalanceController

danfinlay committed:
Make blockTracker an independent param

danfinlay committed:
Merge branch 'master' into AddBalanceController

danfinlay committed:
Add mock random value generator

danfinlay committed:
Make encryptor configurable for keyring-controller

danfinlay committed:
Add computed balance to mock state

kumavis commented:
Changes look good! now just need to get the tests passing -- do they pass locally?

flyswatter commented:
No, tests not passing, I need to learn how to debug with this new integration test runner we have.

danfinlay committed:
Pass blocktracker to balances controller
